src/engine/Preprocessor.cpp

```diff
@@ -781,6 +794,278 @@ void Preprocessor::dumpAllBounds( const String &message )
     printf( "\n" );
 }
 
+bool Preprocessor::constructNetworkLevelReasoner()
+{
+    /*
+      Data structures that we will need for mapping out the DNN
+    */
+    struct Index
+    {
+        Index()
+            : _layer( 0 )
+            , _neuron( 0 )
+        {
+        }
+
+        Index( unsigned layer, unsigned neuron )
+            : _layer( layer )
+            , _neuron( neuron )
+        {
+        }
+
+        bool operator<( const Index &other ) const
+        {
+            if ( _layer < other._layer )
+                return true;
+            if ( _layer > other._layer )
+                return false;
+
+            return _neuron < other._neuron;
+        }
+
+        unsigned _layer;
+        unsigned _neuron;
+    };
+
+    struct NeuronInformation
+    {
+        NeuronInformation()
+            : _weightedSumEquation( NULL )
+            , _activationFunction( NULL )
+        {
+        }
+
+        unsigned _weightedSumVariable;
+        unsigned _activationVariable;
+
+        PiecewiseLinearFunctionType _activationType;
+
+        const Equation *_weightedSumEquation;
+        const PiecewiseLinearConstraint *_activationFunction;
+    };
+
+    Map<Index, NeuronInformation> indexToNeuron;
+    Map<unsigned, Index> variableToIndex;
+    Map<unsigned, unsigned> layerToLayerSize;
+
+    ASSERT( !_preprocessed._networkLevelReasoner );
+
+    const List<Equation> &equations( _preprocessed.getEquations() );
+    const List<PiecewiseLinearConstraint *> plConstraints( _preprocessed.getPiecewiseLinearConstraints() );
+
+    // If the activation functions are not yet supported by the NLR,
+    // abort
+    for ( const auto constraint : plConstraints )
+        if ( !NetworkLevelReasoner::functionTypeSupported( constraint->getType() ) )
+            return false;
+
+    // Attempt to figure out the layered structure
+    List<unsigned> inputLayer;
+    List<unsigned> previousLayer;
+    List<unsigned> currentLayer;
+    List<unsigned> outputLayer;
+    unsigned currentLayerIndex = 1;
+
+    // The input and output layer should be defined in the input query
+    inputLayer = _preprocessed.getInputVariables();
+    previousLayer = inputLayer;
+    unsigned numVariablesHandledSoFar = inputLayer.size();
+    outputLayer = _preprocessed.getOutputVariables();
+    layerToLayerSize[0] = inputLayer.size();
+
+    // Store the input neurons in the topology
+    unsigned count = 0;
+    for ( const auto &neuron : inputLayer )
```

Comment:
Activation function will be unitialized, maybe add Identity to PiecewiseLinearFunctionType to denote input layer nodes? (in case we ever need to distinguish nodes on the fly, without the topology)

Reply:
This is currently compatible with the NLR that is constructed through the ACAS parser. I agree that we could add the identity function, but will leave this for the future.